aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -473,6 +474,38 @@ static Tensor& prod_out_impl(Tensor& result, const Tensor& self, IntArrayRef dim
   return result;
 }
 
+Tensor trace_cpu(const Tensor& self) {
```

Comment:
nit: The best place to put this might be TriangularOps.cpp (trace_cuda is in TriangularOps.cu). Alternatively we can move trace_cuda out of TriangularOps.cu for consistency

Reply:
the problem is you need the specialized promotion logic from here.  In theory you could split things up but this doesn't seem worth it given the promotion logic is really specialized for reductions.